### PR TITLE
Task_10. Get read and write tokens

### PR DESCRIPTION
**Pull request by AllaKashevarova**

### CODE REVIEW

File: .idea/encodings.xml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="Encoding">
+    <file url="file://$PROJECT_DIR$/src/main/java" charset="UTF-8" />
+    <file url="file://$PROJECT_DIR$/src/main/resources" charset="UTF-8" />
+  </component>
```
Comment (KirKarpov) [inline .idea/encodings.xml:6]: Everything from .idea folder should be added to gitignore.

File: src/main/java/com.coherent/BaseClass.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+public class BaseClass {
+    CloseableHttpClient httpClient;
+
+    //TODO 1. Class HttpClient - DONE
+    //TODO 2. Class with various methods (e.g. SendPost) - DONE
+    //TODO 3. Service logic (authorization e.g. class GetToken) - use 2 layers above - DONE
+    //TODO 4. Save token to some variable to reuse it
+    //TODO 5. Scope: write and read token (think how to organize them in classes)
+    //TODO 6. Parse response (See Class ObjectMapper!) - check in google: "objectmapper baeldung"
```
Comment (KirKarpov) [inline src/main/java/com.coherent/BaseClass.java]: Do you need this class ?

File: src/main/java/com.coherent/HttpClientFactory.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+import java.net.URISyntaxException;
+
+public class HttpClientFactory {
+
+    public CloseableHttpClient createClient() {
+        CredentialsProvider credsProvider = new BasicCredentialsProvider();
+
+        credsProvider.setCredentials(
+                new AuthScope("localhost", 8050),
+                //TODO get creds from authCreds.properties
```
Comment (KirKarpov) [inline src/main/java/com.coherent/HttpClientFactory.java]: Lets do it :) 
Credentials should be in property file

File: src/main/java/com/coherent/token/TokenUrlBuilder.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
+package com.coherent.token;
+
+import org.apache.http.client.utils.URIBuilder;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+
+public class TokenUrlBuilder {
+
+    public static URI buildTokenUri(String scope) {
```
Comment (KirKarpov) [inline src/main/java/com/coherent/token/TokenUrlBuilder.java]: Please move method to HttpClientFactory class
This class can be removed

File: src/main/java/com/coherent/token/ReadToken.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
+package com.coherent.token;
+
+import com.coherent.HttpClientFactory;
+import helpers.JsonParser;
+import org.apache.http.impl.client.CloseableHttpClient;
+
+public class ReadToken {
```
Comment (KirKarpov) [inline src/main/java/com/coherent/token/ReadToken.java]: Please check, but looks like this one isn't needed

File: src/main/java/com/coherent/token/ReadTokenManager.java [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    }
+
+    public static ReadTokenManager getInstance() {
+        if (instance == null) {
+            instance = new ReadTokenManager();
+        }
+        return instance;
+    }
+
+    public void setReadToken(String token) {
```
Comment (KirKarpov) [inline src/main/java/com/coherent/token/ReadTokenManager.java]: Set token method shouldn't be in Singleton pattern

File: src/main/java/com/coherent/token/ReadTokenManager.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
+package com.coherent.token;
+
+public class ReadTokenManager {
```
Comment (KirKarpov) [inline src/main/java/com/coherent/token/ReadTokenManager.java]: Please review implementation of Singleton pattern. Looks like something is missed

File: src/main/java/helpers/JsonParser.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+        JsonNode jsonNode = null;
+        try {
+            jsonNode = objectMapper.readTree(response);
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException(e);
+        }
+        String token = jsonNode.get("access_token").asText();
+
+        return token;
+    }
```
Comment (KirKarpov) [inline src/main/java/helpers/JsonParser.java]: Could you please refactor this one and use objectMapper.readValue(json, TokenResponse.class)
You will need to create TokenResponse class which will have fields from the response that you get when send request to get token.
json - it is your response from backend. You are getting it in WriteTokenReceiver on line 35 : responseBody = EntityUtils.toString(response.getEntity());

File: src/main/java/helpers/JsonParser.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+//    public void writeToFile(ReadTokenManager token) {
+//        try (FileWriter writer = new FileWriter("src/main/resources/" + cart.getCartName() + ".json")) {
+//            writer.write(gson.toJson(cart));
+//        } catch (IOException e) {
+//            e.printStackTrace();
+//        }
+//    }
+
+
+    public String readFromJson (String response) {
```
Comment (KirKarpov) [inline src/main/java/helpers/JsonParser.java]: please make it static

File: src/main/java/org/example/Main.java [hunk 10]
```diff
@@ -1,7 +1,39 @@
...
+        System.out.println(readToken);
+
+        JsonParser jsonParser = new JsonParser();
+        String accessToken = jsonParser.readFromJson(readToken);
+        System.out.println(accessToken);
+
+        ReadToken readToken1 = new ReadToken();
+        String parsedReadToken = readToken1.extractReadToken();
+        ReadTokenManager readTokenManager = new ReadTokenManager();
+        readTokenManager.setReadToken(parsedReadToken);
```
Comment (KirKarpov) [inline src/main/java/org/example/Main.java]: It will not work in such way :)

File: src/main/java/helpers/PropertiesHelper.java [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+import java.util.Properties;
+
+public class PropertiesHelper {
+
+    public String propertiesReader(String prop, String propFileName) {
+        Properties properties = new Properties();
+        InputStream inputStream;
+
+        File file = new File("src/main/resources/authCreds.properties" + propFileName);
+        try {
```
Comment (KirKarpov) [inline src/main/java/helpers/PropertiesHelper.java]: try (InputStream inputStream=new FileInputStream(file); ) {
}

File: .idea/encodings.xml (hunk 1 above)
Comment (AllaKashevarova) [reply .idea/encodings.xml:6]: Resolved

File: src/main/java/com/coherent/token/TokenUrlBuilder.java (hunk 4 above)
Comment (AllaKashevarova) [reply src/main/java/com/coherent/token/TokenUrlBuilder.java]: Resolved

File: src/main/java/com/coherent/token/ReadTokenManager.java (hunk 6 above)
Comment (AllaKashevarova) [reply src/main/java/com/coherent/token/ReadTokenManager.java]: Removed

File: src/main/java/com.coherent/BaseClass.java (hunk 2 above)
Comment (AllaKashevarova) [reply src/main/java/com.coherent/BaseClass.java]: No. Removed

File: src/main/java/com/coherent/token/ReadToken.java (hunk 5 above)
Comment (AllaKashevarova) [reply src/main/java/com/coherent/token/ReadToken.java]: Removed

File: src/main/java/com/coherent/token/ReadTokenManager.java (hunk 7 above)
Comment (AllaKashevarova) [reply src/main/java/com/coherent/token/ReadTokenManager.java]: Refactored

File: src/main/java/helpers/JsonParser.java (hunk 8 above)
Comment (AllaKashevarova) [reply src/main/java/helpers/JsonParser.java]: Refactored as discussed on a meeting. I deleted the whole JsonParser class and added deserialization logic to the ReadTokenReceiver and WriteTokenReceiver class. I used ObjectMapper class for that.

File: src/main/java/helpers/JsonParser.java (hunk 9 above)
Comment (AllaKashevarova) [reply src/main/java/helpers/JsonParser.java]: N/A since the class has been removed

File: src/main/java/org/example/Main.java (hunk 10 above)
Comment (AllaKashevarova) [reply src/main/java/org/example/Main.java]: Got it :) already fixed

File: src/main/java/helpers/PropertiesHelper.java (hunk 11 above)
Comment (AllaKashevarova) [reply src/main/java/helpers/PropertiesHelper.java]: Could I clarify why is that? I already have a declared variable:

InputStream inputStream;

File: src/main/java/com.coherent/HttpClientFactory.java (hunk 3 above)
Comment (AllaKashevarova) [reply src/main/java/com.coherent/HttpClientFactory.java]: Done

File: .idea/encodings.xml (hunk 1 above)
Comment (KirKarpov) [reply .idea/encodings.xml:6]: I don't see gitignore file changes :)

File: src/main/java/com/coherent/token/WriteTokenReceiver.java [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+            String responseBody = EntityUtils.toString(response.getEntity());
+            ObjectMapper objectMapper = new ObjectMapper();
+            JsonNode jsonNode = objectMapper.readTree(responseBody);
+            parsedWriteToken = jsonNode.get("access_token").asText();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+        return parsedWriteToken;
+
+    }
```
Comment (KirKarpov) [inline src/main/java/com/coherent/token/WriteTokenReceiver.java]: Code duplication. The code mostly the same as in ReadTokenReceiver class. Please think how you can reuse code

File: src/main/java/com/coherent/token/WriteTokenManager.java [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+            writeTokenValue = writeTokenReceiver.getWriteToken();
+
+            instance = new WriteTokenManager(writeTokenValue);
+        }
+        return instance;
+    }
+
+    public String getWriteToken(){
+        return writeTokenValue;
+    }
```
Comment (KirKarpov) [inline src/main/java/com/coherent/token/WriteTokenManager.java]: Code mostly the same as for ReadTokenManager. You can left it as is or think on solution how to reuse code

File: src/main/java/helpers/PropertiesHelper.java (hunk 11 above)
Comment (KirKarpov) [reply src/main/java/helpers/PropertiesHelper.java]: It is just about to clean up a bit code. Such construction called try with resources which allows you to reduce a bit code. More about it :https://www.baeldung.com/java-try-with-resources

**Review by KirKarpov — CHANGES_REQUESTED** [review]

File: .idea/encodings.xml (hunk 1 above)
Comment (AllaKashevarova) [reply .idea/encodings.xml:6]: oh, maybe I've missed the slash symbol. I added this to .gitignore: 
.idea/

Is that the correct line?

File: src/main/java/helpers/PropertiesHelper.java (hunk 11 above)
Comment (AllaKashevarova) [reply src/main/java/helpers/PropertiesHelper.java]: Got it! Great tip, thanks! Refactored

File: src/main/java/com/coherent/token/WriteTokenReceiver.java (hunk 12 above)
Comment (AllaKashevarova) [reply src/main/java/com/coherent/token/WriteTokenReceiver.java]: Done. I created a TokenExtractor class and reused its method getToken() in WriteTokenReceiver and ReadTokenReceiver classes

File: src/main/java/com/coherent/token/WriteTokenManager.java (hunk 13 above)
Comment (AllaKashevarova) [reply src/main/java/com/coherent/token/WriteTokenManager.java]: Refactored

File: .idea/encodings.xml (hunk 1 above)
Comment (KirKarpov) [reply .idea/encodings.xml:6]: should be good now

**Review by KirKarpov — APPROVED** [review]

**Pull request closed without merging**